### ISSUE TITLE
Query Variable Options Deduped by Value instead of Key

### ISSUE THREAD

**Issue by carlverge** [grafana/grafana#67077]
## What happened:

In the Grafana UI, query variables options are deduplicated by their `value` instead of `text`. Effectively, it becomes like a map/dict being dedup'd by the key instead of the value.

<img width="510" alt="Screenshot 2023-04-22 at 9 35 34 AM" src="https://user-images.githubusercontent.com/1541056/233796203-750d3cc0-b662-4c3f-b19b-1cf0fb1c1dcb.png">

## What you expected to happen:

Query variable options are deduplicated based on the text field (the dropdown key) rather than the value.

In the repro example below, the output options are `dupkey, dupkey, dupval-a`. I would expect them to be `dupkey, dupval-a, dupval-b`.

## How to reproduce it (as minimally and precisely as possible):

Make a query variable with a MySQL datasource and the following query:
```sql
SELECT * FROM (
  SELECT 'dupkey=dupkey-a'
  UNION ALL SELECT 'dupkey=dupkey-b'
  UNION ALL SELECT 'dupval-a=dupval'
  UNION ALL SELECT 'dupval-b=dupval'
)t
```

Then use the following regex:
```
(?<text>.*)=(?<value>.*)
```

## Anything else we need to know?:

It looks like the relevant line of code to change is: https://github.com/grafana/grafana/blob/afc9925dbfa55886cc85a9a2106ed9892d7df239/public/app/features/variables/query/reducer.ts#L134

## Environment:
- Grafana version: 9.5.0 (also 9.1.6)
- Data source type & version:
- OS Grafana is installed on:
- User OS & Browser:
- Grafana plugins:
- Others:

**Comment by mdvictor** [comment]
Hey @carlverge. Thank you for opening this issue! We will further investigate and try to bring a resolution as soon as possible.

**Comment by jHumeLPL** [comment]
Yes, this is continually causing problems for us.  It wouldn't be a big deal except the deduping seems to be happening before a plugin's parsing can even get to it.  

For example, if I am using CSV data in the Infinity plugin and I try to filter the data using the Backend or UQL filters, I can't filter on fields that are deduped by the variable.  Which is weird because you would think that the plugin parsing features could be used to narrow down the results and _then_ that would be fed to the variable engine where the dedupe would occur, but that doesn't seem to be the case.

It seems like, at least with the Infinity plugin, that the columns are selected, then sent to the variable engine in Grafana where they get deduped, and then _back_ to the Infinity parsing engine where I can then filter the now deduped results, which is useless.  If I could just filter the entire results _before_ sending to the variable engine to have them deduped, there wouldn't even be any duplicate results to be processed anyways, and I wouldn't care.

Ideally, it seems like a toggle could just be added to say "Remove Duplicates".  Let us choose whether we want the duplicate results removed.  Either that or at least have the deduplication done after the plugin has fully selected and processed the results.

**Comment by gabor** [comment]
the mentioned deduplication is happening in the dashboard-variable-handling code, so i do not think the problem is related to the mysql datasource plugin, reassigning this to the related people.

feel free to assign it back if mysql-plugin-changes are needed 👍 .

**Comment by carl-db** [comment]
Correct, this is definitely not the MySQL datasource. I have linked the relevant line of code causing the problem at the bottom of the issue description.

**Comment by kaydelaney** [comment]
Hi, wanted to start looking into this but I'm a little confused.
The title of the issue is "Query Variable Options Deduped by Value instead of Key", but in the body of the issue it's mentioned that
> Effectively, it becomes like a map/dict being dedup'd by the key instead of the value.

Personally the behavior as it is seems correct to me. In your example, the preview of values shows "dupkey, dupkey, dupval-a", and while admittedly having two "dupkey" labels is confusing, the corresponding values are "dupkey-a, dupkey-b, dupval", which makes sense to me, as all values are represented and can be chosen.
If it were the other way around, and we de-duplicated by key and had "dupkey, dupval-a, dupval-b" as you mentioned, the corresponding values would be "dupkey-a, dupval, dupval", meaning that "dupval-b" could no longer be chosen, and two of the values would be identical.
Let me know if I'm misunderstanding something here. Thanks.

**Comment by harisrozajac** [comment]
@jHumeLPL I agree with @kaydelaney 's comment above as I'm confused too. We are deduplicating by value as expected. Let us know if we are misunderstanding.
